Python/ceval.c

```diff
@@ -3684,6 +3686,104 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
+            }
+        }
+
+        TARGET(COMPARE_OP_FLOAT) {
+            SpecializedCacheEntry *caches = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &caches[0].adaptive;
+            PyObject *right = TOP();
+            PyObject *left = SECOND();
+            DEOPT_IF(!PyFloat_CheckExact(left), COMPARE_OP);
+            DEOPT_IF(!PyFloat_CheckExact(right), COMPARE_OP);
+            STAT_INC(COMPARE_OP, hit);
+            double dleft = PyFloat_AS_DOUBLE(left);
+            double dright = PyFloat_AS_DOUBLE(right);
+            int cmp;
+            switch (cache0->original_oparg) {
+                case Py_LT: cmp = (dleft < dright); break;
+                case Py_LE: cmp = (dleft <= dright); break;
+                case Py_EQ: cmp = (dleft == dright); break;
+                case Py_NE: cmp = (dleft != dright); break;
+                case Py_GT: cmp = (dleft > dright); break;
+                case Py_GE: cmp = (dleft >= dright); break;
+                default: Py_UNREACHABLE();
+            }
+            // This cannot fail
+            PyObject *res = PyBool_FromLong(cmp);
+            assert(!PyErr_Occurred());
+            SET_SECOND(res);
+            STACK_SHRINK(1);
+            Py_DECREF(left);
+            Py_DECREF(right);
+            PREDICT(POP_JUMP_IF_FALSE);
+            PREDICT(POP_JUMP_IF_TRUE);
+            DISPATCH();
+        }
+
+        TARGET(COMPARE_OP_INT) {
+            SpecializedCacheEntry *caches = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &caches[0].adaptive;
+            PyObject *right = TOP();
+            PyObject *left = SECOND();
+            DEOPT_IF(!PyLong_CheckExact(left), COMPARE_OP);
+            DEOPT_IF(!PyLong_CheckExact(right), COMPARE_OP);
+            STAT_INC(COMPARE_OP, hit);
+            // This cannot fail.
+            PyObject *res = _PyLong_RichCompare(
+                (PyLongObject *)left, (PyLongObject *)right,
+                cache0->original_oparg);
+            assert(res != NULL);
+            assert(!PyErr_Occurred());
+            SET_SECOND(res);
+            STACK_SHRINK(1);
+            Py_DECREF(left);
+            Py_DECREF(right);
+            PREDICT(POP_JUMP_IF_FALSE);
+            PREDICT(POP_JUMP_IF_TRUE);
+            DISPATCH();
+        }
+
+        TARGET(COMPARE_OP_STR) {
+            SpecializedCacheEntry *caches = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &caches[0].adaptive;
+            PyObject *right = TOP();
+            PyObject *left = SECOND();
+            DEOPT_IF(!PyUnicode_CheckExact(left), COMPARE_OP);
+            DEOPT_IF(!PyUnicode_CheckExact(right), COMPARE_OP);
+            DEOPT_IF(!PyUnicode_IS_READY(left), COMPARE_OP);
+            DEOPT_IF(!PyUnicode_IS_READY(right), COMPARE_OP);
+            STAT_INC(COMPARE_OP, hit);
+            assert(cache0->original_oparg == Py_EQ || cache0->original_oparg == Py_NE);
+            int cmp = Py_Is(left, right) || _PyUnicode_EQ(left, right);
```

Comment:
I don't think `_PyUnicode_EQ` is correct here. The docs say that a hash equality check should already have been performed.

Reply:
Using PyUnicode_RichCompare would call PyUnicode_READY on its arguments, which should almost always just check that PyUnicode_IS_READY. PyUnicode_RichCompare then passes to unicode_compare_eq, while _PyUnicode_EQ passes to unicode_eq in stringlib/eq.h. The implementations of unicode_eq and unicode_compare_eq are essentially the same, but _PyUnicode_EQ was already exposed, so I went with that.

Is the concern the different branching characteristics of this versus the very-likely-equal hash table case? Other than that, the function seems to do exactly what you would expect.

Reply:
The docs say that a hash equality should already have been performed before using `_PyUnicode_EQ`, but you haven't done that.

So either the docs are wrong, or you code is wrong. From what you say, it sounds like the docs are wrong, and `_PyUnicode_EQ` can be used on any readied strings.

This is moot, of course, if you make the change discussed above.